Vignettes (#1545)
insightsengineering/teal: vignettes/managing-reproducible-report-documents-in-teal.rmd

```{r custom_reporter}
library(teal.reporter)

# Create a custom reporter with specific configuration
custom_reporter <- Reporter$new()
custom_card <- teal_card("## Welome", "Welcome message", head(iris))
custom_reporter$append_cards(list(WelcomeCard = custom_card))

# Use the custom reporter in your app
app_custom <- init(
  data = teal_data(IRIS = iris, MTCARS = mtcars),
  modules = modules(
    example_module("Analysis 1"),
    example_module("Analysis 2")
  ),
  reporter = custom_reporter                        # TODO: doesn't work yet, app do not start with a predefined card
)
# Start app with
# if (interactive()) shinyApp(app_custom$ui, app_custom$server)
```

##### 4. Custom Reporter with a template

Custom initial reporter can also include a template function that will be applied to every card:

```{r custom_reporter_template}
library(teal.reporter)

# Create a custom reporter with specific configuration
custom_reporter_template <- Reporter$new()
custom_reporter_template$set_template(
  # Need to return named `teal_card`
  function(card) {
    tcard <- c("Templated information", card)
    attributes(tcard) <-  attributes(card)

>     tcard

Author: This is nasty. `c` removes all classes and attributes. We can substitute this example to use `edit_teal_card`

Reviewer: You should use `append(card, <something to add>, after = 0)` - then it doesn't loose a class because a first argument is a `card`

Author: Yes, I think I'll go with this one

---
title: "Managing Reproducible Report Documents in teal"
author: "NEST CoreDev"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Managing Reproducible Report Documents in teal}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

> `teal` provides comprehensive reporting capabilities that allow users to generate reproducible documents from their analysis modules. This vignette covers the key components for managing report documents in `teal` applications.

Reviewer: ```suggestion
`teal` provides reporting capabilities that allow app users to compose reproducible documents from their analysis modules. This vignette covers the key components for managing report documents in `teal` applications.
```

> The reporting system offers three levels of control:
> 
> 1. **Global reporting control** via the `reporter` parameter in `init()`
> 2. **Module-level control** using `disable_report()` to exclude specific modules  
> 3. **Report customization** with the `after` function to enhance report content

Reviewer: Three levels suggests that you go deeper or higher in the hierarchy. We operate here only on global/module (local) level.
```suggestion
The reporting system offers two types of control:

1. **Global reporting control** via the `reporter` parameter in `init()`
2. **Module-level control** via `after()` and `disable_report()`
```

Author: Make sense, thanks!

This vignette demonstrates how to use these features effectively to create flexible and customized reporting workflows.

> ## Global and Local Reporting Control

Reviewer: I think it's worth to split Global and Local according to the bullet points from above

```suggestion
## Global Reporting Control
```

##### 3. Custom Reporter 
For advanced use cases, you can provide a custom initial reporter object:

```{r custom_reporter}
library(teal.reporter)

# Create a custom reporter with specific configuration
custom_reporter <- Reporter$new()
custom_card <- teal_card("## Welome", "Welcome message", head(iris))
metadata(custom_card, "title") <- "The title"
custom_reporter$append_cards(list(WelcomeCard = custom_card))

# Use the custom reporter in your app
app_custom <- init(
  data = teal_data(IRIS = iris, MTCARS = mtcars),
  modules = modules(
    example_module("Analysis 1"),
    example_module("Analysis 2")
  ),
  reporter = custom_reporter                        # TODO: doesn't work yet, app do not start with a predefined card
)
# Start app with
# if (interactive()) shinyApp(app_custom$ui, app_custom$server)
```

##### 4. Custom Reporter with a template

Custom initial reporter can also include a template function that will be applied to every card:

```{r custom_reporter_template}
library(teal.reporter)

# Create a custom reporter with specific configuration
custom_reporter_template <- Reporter$new()
custom_reporter_template$set_template(
  # Need to return named `teal_card`

>   function(card) {

Reviewer: Can you show some practical information, like setting a Rmd template or adding some header/footer etc.?

Author: I was rather thinking about setting some templated text for disclaimers, data sources, data privacy information, etc as an information added on top or at the bottom of the card. This sentence can also be added to this example.

The `Rmd` template could be useful, if we create some text that can be interpreted as `rmarkdown`. Maybe including a logo that is taken from the web?

    tcard <- c("Templated information", card)
    attributes(tcard) <-  attributes(card)
    tcard
  }
)

# Use the custom reporter in your app
app_custom <- init(
  data = teal_data(IRIS = iris, MTCARS = mtcars),
  modules = modules(
    example_module("Analysis 1"),
    example_module("Analysis 2")
  ),
  reporter = custom_reporter_template
)
# Start app with
# if (interactive()) shinyApp(app_custom$ui, app_custom$server)
```

#### Module-Level Reporting Control

> While the `reporter` parameter controls reporting globally, you can also disable reporting for specific modules using `disable_report()`. This allows you to have fine-grained control over which modules support reporting:

Reviewer: Both, `after` and `disable_report` can control module's report. After is more powerful and it doesn't restrict to the report only but to the whole `teal_report/teal_data` object. We can showcase how to modify the report using `after` 
 
```suggestion
While the `reporter` parameter controls reporting globally, you can also manage a report document for a specific module using `after()` `disable_report()`:
```

Author: @gogonzo there is a section following this sentence that presents the usage of `after`:
`Using the `after` Function to Customize Module Report Cards`

Do you think it's enough? It shows that you can adjust the `teal_card` of `teal_report`, but we can apply `join_keys` modifications as well, and also an `eval_code` to show that the whole `teal_data/teal_report` can be changed in any direction, not only the `@teal_report` field.
